flink-ml-lib/src/test/java/org/apache/flink/ml/classification/OnlineLogisticRegressionTest.java

```diff
@@ -0,0 +1,444 @@
+                        Vectors.dense(0.43782349911420193, 0.5621765008857981),
+                        0L);
+        DataStream<Row> results =
+                tEnv.toDataStream(onlineLogisticRegressionModel.transform(onlinePredictTable)[0]);
+        List<Row> resultList = IteratorUtils.toList(results.executeAndCollect());
+        Assert.assertEquals(expected, resultList.get(0));
+    }
+
+    private static void verifyPredictionResult(Table output, int expectedNum) throws Exception {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) output).getTableEnvironment();
+        DataStream<Row> stream = tEnv.toDataStream(output);
+        List<Row> result = IteratorUtils.toList(stream.executeAndCollect());
+        Map<Long, Tuple2<Double, Double>> correctRatio = new HashMap<>();
+
+        for (Row row : result) {
+            long modelVersion = row.getFieldAs(MODEL_VERSION_COL);
+            Double pred = row.getFieldAs(PREDICT_COL);
+            Double label = row.getFieldAs(LABEL_COL);
+            if (correctRatio.containsKey(modelVersion)) {
+                Tuple2<Double, Double> t2 = correctRatio.get(modelVersion);
+                if (pred.equals(label)) {
+                    t2.f0 += 1.0;
+                }
+                t2.f1 += 1.0;
+            } else {
+                correctRatio.put(modelVersion, Tuple2.of(pred.equals(label) ? 1.0 : 0.0, 1.0));
+            }
+        }
+        int numModel = 0;
+        for (Long id : correctRatio.keySet()) {
+            assertEquals(1.0, correctRatio.get(id).f0 / correctRatio.get(id).f1, 1.0e-5);
+            numModel++;
+        }
+        if (expectedNum != -1) {
+            assertEquals(expectedNum, numModel);
+        }
+    }
+
+    /** Random selects samples with fixed size from a given sample list. */
+    private static class RandomSample implements SourceFunction<Row> {
+        private volatile boolean isRunning = true;
+        private final long timeInterval;
+        private final long numSample;
+        private final List<Row> data;
+
+        public RandomSample(long timeInterval, long numSample, List<Row> data) {
+            this.timeInterval = timeInterval;
+            this.numSample = numSample;
+            this.data = data;
+        }
+
+        @Override
+        public void run(SourceContext<Row> ctx) throws Exception {
+            int size = data.size();
+            for (int i = 0; i < numSample; ++i) {
+                int idx = i % size;
+                if (isRunning) {
+                    ctx.collect(data.get(idx));
+                    if (timeInterval > 0) {
+                        Thread.sleep(timeInterval);
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void cancel() {
+            isRunning = false;
+        }
+    }
+
+    private Table getTable(
+            int timeInterval, int numSample, List<Row> data, int parallel, boolean isSparse) {
```

Comment:
Would it be more readable to rename `parallel` as `sourceParallelism`?

Would it be more readable to rename this method as `getSourceTable`?

When the parallelism is 2, the last 50% of elements of `data` will never be emitted by the source function. Is this expected?

Would you help explain the motivation to test source function with different parallelism? For example, would it be sufficient to only test `sourceParallelism = 2`?

Reply:
For the first question, I don't think so. The data is random sampling from a fixed dataset.


The user set the total num samples, we need to calculate the num samples on every worker, then we need source parallel parameter.